### PR TITLE
Create KerasModel

### PR DESCRIPTION
**Pull request by peastman**
This is the second piece of #1461.  There's still a lot to do, but I'm posting it now to get feedback.  As much as possible I'm trying to keep the API consistent with TensorGraph, but some differences seem to be unavoidable:

1. Each TensorGraph creates its own tf.Graph and uses that for all its calculations.  Making KerasModel work the same way seems to be impractical.  Many layers create TensorFlow objects in their constructors, which are called by the user before they've even created the KerasModel.  So instead I'm leaving everything in the default graph, which is how Keras expects things to work.

2. It's much stricter about the shapes and types of inputs.  TensorGraph knows the expected shape and dtype of every input, label, and weight, so it can automatically do a lot of conversions.  For example, if the labels are supposed to have shape (batch_size, 1) but instead you pass in a 1D array of shape (batch_size,), it automatically adds the extra axis for you.  That isn't possible with a Keras model, because we don't know in advance what shape the labels are supposed to have.  The shapes of most inputs are determined from the values you pass in the first time the model is invoked.  If you pass an array of shape (batch_size,), it builds a tensor of that shape.  If the loss function then tries to compare the labels of shape (batch_size,) to outputs of shape (batch_size, 1), you'll get an exception.

3. The functions that take generators instead of datasets work a bit differently.  In TensorGraph, the generator produces dicts that map layer objects to NumPy arrays.  In KerasModel, it produces tuples of the form (inputs, labels, weights).

### CODE REVIEW

**Comment by peastman** [comment]
My latest changes make it a lot simpler to define the loss function, and also try to make it a little more tolerant regarding shapes of arrays and tensors.

**Comment by rbharath** [comment]
Took a quick skim. This is looking really nice so far! Will try to do a more thorough pass later

**Comment by peastman** [comment]
I think I'm approaching a complete version.  It supports most features of TensorGraph, but there are a few missing.

- I didn't bother to implement `make_estimator()`.  My impression is that estimators are on their way out now that Google has decided to standardize on Keras models.  And anyway, you can already create an estimator from a tf.keras.Model.

- I also haven't implemented submodels.  Keras has its own mechanisms for achieving the same goals (by composition of models).  Perhaps in the future I'll change my mind and implement them, but for the moment I'd like to see if we can just use the standard mechanisms.

- There's no `save()` method.  The problem is that the loss can be specified by an arbitrary Python function, and there's no way to save that.  tf.keras.Model has its own mechanisms for serialization, so you can still save the model that's being wrapped, but you'll need to provide the code to wrap it in a KerasModel object.

- TensorBoard support is a bit limited at the moment: the only thing it logs is the loss.  Feedback on what else to include would be helpful.

Does that sound reasonable?

**Comment by rbharath** [comment]
This set of tradeoffs sounds pretty reasonable to me. Leaning on the standard Keras way of doing things seems like a safe bet

**Comment by peastman** [comment]
I'm very confused.  One test case is consistently failing on travis, but only on Python 3.5, not 2.7.  Outside travis it never fails.  I've tried running it on two different computers, with and without GPUs, under Python 3.5, 3.6, and 3.7.  I've never gotten a single failure.  But on travis it fails every time.  But again only on Python 3.5.

**Comment by peastman** [comment]
So I added some debugging code to get more information about why it's failing.  And what happens when I do that?  Of course, it passes!

Grrrr.

Try running it again...

**Comment by peastman** [comment]
Ok!  I think this is ready for you to review and merge if you're happy with it.
